src/jobflow/schemas/job_store.py

```diff
@@ -0,0 +1,39 @@
+"""A Pydantic model for Jobstore document."""
+
+import typing
+
+from pydantic import BaseModel, Field
+
+from jobflow.core.reference import OutputReference
+
+
+class JobStoreDocument(BaseModel):
+    """A Pydantic model for Jobstore document."""
+
+    uuid: str = Field(
+        None, description="A unique identifier for the job. Generated automatically."
+    )
+    index: int = Field(
+        None,
+        description="The index of the job (number of times the job has been replaced.",
+    )
+    output: typing.Any = Field(
```

Comment:
Instead of `typing.Any`, this could be made a Generic:

```
T = TypeVar("T")

class JobStoreDoc(BaseModel, Generic[T]):

    ...

    output: T
```

This would allow people to easily sub-class when they're working with a JobStore document with a specific output type.

Reply:
On this note, I had a similar document model implemented that I had intended to upstream. One addition I found helpful was:

```
    @validator("output", pre=True)
    def reserialize_output(cls, v):
        if isinstance(v, dict) and "@module" in v and "@class" in v:
            v = MontyDecoder().process_decoded(v)
        return v
```

Note this is in pydantic v1 syntax, I think it will have to be modified slightly for pydantic v2.

This is useful in the scenario that you're loading a dict directly, e.g. `JobFlowDocument(**dict)`, rather than using MontyDecoder which I think would take care of it automatically.

Reply:
Thanks Matt! These are good suggestions. We have another PR that this one depends on that isolates the implementation of that model right here: https://github.com/materialsproject/jobflow/pull/424. This one is a clunky use of git.

@hrushikesh-s Do you want to update your PR with these suggestions?

Reply:
@mkhorton thanks for the suggestions. 
@mcgalcode , yes I'll update my PR accordingly.

Reply:
@mkhorton: Jobflow uses Pydantic v1 (see [here](https://github.com/materialsproject/jobflow/blob/main/pyproject.toml#L34)). Is there any plan to upgrade it to v2 anytime soon? If not then your implementation of `@validator("output", pre=True)` should work as it is, and we don't need to change the syntax. 
Maybe I'm missing out on something there?

Reply:
I plan to migrate Jobflow to pydantic2 very very soon. I am waiting on Jason to patch emmet and maggma first, which he said should be done in a day or two.

Reply:
@Andrew-S-Rosen thanks for the update. For the time being, I've implemented the `validator` as per `pydantic v1` syntax in #424, but changing it to `pydantic v2` should be straightforward. I'll modify it once your migration from `pydantic v1` to `v2` is complete.

Reply:
Sounds good! Thanks. Will be good to keep tabs on what needs switching and where.